### ISSUE TITLE
The Content-Length in header will disappear while retrying S3 pubObject with a 0 size file.

### ISSUE THREAD

**Issue by weichenglee** [aws/aws-sdk-php#497]
While retrying S3 pubObject with a 0 size file, the Content-Length in header will disappear and will get getMissingContentLengthException.

agent_server_s3.php:706:_main: exec putObject({"Body":"\/volume1\/zero_files\/50230","Bucket":"bat-wadeli","Key":"BAT_Wolf_1\/zero_files\/50230","StorageClass":"STANDARD"})
> PUT /BAT_Wolf_1/zero_files/50230 HTTP/1.1
Host: bat-wadeli.s3.hicloud.net.tw
x-amz-storage-class: STANDARD
User-Agent: aws-sdk-php2/2.7.20 Guzzle/3.9.2 curl/7.36.0 PHP/5.5.21
Date: Thu, 26 Feb 2015 06:41:34 +0000
Authorization: XXXXX
Content-Length: 0

< HTTP/1.0 500 Internal Server Error
< x-amz-request-id: DWCHBCDPRPGS9RVL
< x-amz-id-2:
< Content-Type: application/xml
< Date: Thu, 26 Feb 2015 06:46:34 GMT
< Connection: close
<
> PUT /BAT_Wolf_1/zero_files/50230 HTTP/1.1
Transfer-Encoding: chunked
Host: bat-wadeli.s3.hicloud.net.tw
x-amz-storage-class: STANDARD
User-Agent: aws-sdk-php2/2.7.20 Guzzle/3.9.2 curl/7.36.0 PHP/5.5.21
Date: Thu, 26 Feb 2015 06:46:34 +0000
Authorization: XXXXX

< HTTP/1.0 411 Length Required
< x-amz-request-id: C8GE9VFYZ0MQVX4M
< x-amz-id-2:
< Content-Type: application/xml
< Date: Thu, 26 Feb 2015 06:46:35 GMT
< Connection: close
<
agent_server_s3.php:63:_convert_exception: Aws\S3\Exception\MissingContentLengthException: AWS Error Code: MissingContentLength, Status Code: 411, AWS Request ID: C8GE9VFYZ0MQVX4M, AWS Error Type: client, AWS Error Message: You must provide the Content-Length HTTP header., User-Agent: aws-sdk-php2/2.7.20 Guzzle/3.9.2 curl/7.36.0 PHP/5.5.21

**Comment by mtdowling** [comment]
It's interesting that the `Content-Length: 0` header was removed and a `Transfer-Encoding: chunked` header was added.

I tried reproducing this with the following test case, but was unable:

```php
<?php
require 'vendor/autoload.php';

$client = Aws\S3\S3Client::factory([
    'region' => 'us-east-1'
]);

$history = new \Guzzle\Plugin\History\HistoryPlugin();

$mock = new Guzzle\Plugin\Mock\MockPlugin([
    new \Guzzle\Http\Message\Response(500),
    new \Guzzle\Http\Message\Response(200),
]);

$client->getEventDispatcher()->addSubscriber($mock);
$client->getEventDispatcher()->addSubscriber($history);

$client->putObject([
    'Key'    => 'foo',
    'Bucket' => 'bar',
    'Body'   => ''
]);

echo $history;
```

Does running the above code example include a Content-Length header on the retry request?

**Comment by weichenglee** [comment]
I tried your code and the Content-Length header was included.

However, I tried another test to stress the server and the problem is reproducible. (not AWS)

The results of the two tests are different.

```php
<?php

require './aws.phar';

use Aws\S3\S3Client;

$aws_config = array(
        'key' => "xxxx",
        'secret' => "xxxx",
        'base_url' => "http://s3.hicloud.net.tw",
        'scheme' => "http",
        'curl.options' => array(
                'CURLOPT_VERBOSE' => true,
        ),
);

$client = S3Client::factory($aws_config);

for ($i = 0; $i <= 1000; $i++) {
        $in_json = array(
                "Bucket" => "bat-wadeli",
                "Key" => "zero_file/zero",
                "SourceFile" => "/volume1/zero"
        );
        $res = $client->putObject($in_json);
}

?>
```
and got
```
> PUT /zero_file/zero HTTP/1.1
Host: bat-wadeli.s3.hicloud.net.tw
User-Agent: aws-sdk-php2/2.7.21 Guzzle/3.9.2 curl/7.36.0 PHP/5.5.21
Date: Thu, 12 Mar 2015 06:48:26 +0000
Authorization: AWS SE41NTAxMjQ3OTEzODM2MTMzNDY1NDE:WFJMj4WuessE5DXtGjmVRQQAaLM=
Content-Length: 0

< HTTP/1.0 500 Internal Server Error
< x-amz-request-id: 6RORGA1TVTPN1PUM
< x-amz-id-2:
< Content-Type: application/xml
< Date: Thu, 12 Mar 2015 06:48:27 GMT
< Connection: close
<
> PUT /zero_file/zero HTTP/1.1
Transfer-Encoding: chunked
Host: bat-wadeli.s3.hicloud.net.tw
User-Agent: aws-sdk-php2/2.7.21 Guzzle/3.9.2 curl/7.36.0 PHP/5.5.21
Date: Thu, 12 Mar 2015 06:48:27 +0000
Authorization: AWS SE41NTAxMjQ3OTEzODM2MTMzNDY1NDE:r8z8kfhDHWbN+PWWBSfpjOcr8GE=

< HTTP/1.0 411 Length Required
< x-amz-request-id: ZESZXB7QM931H7F5
< x-amz-id-2:
< Content-Type: application/xml
< Date: Thu, 12 Mar 2015 06:48:28 GMT
< Connection: close
<
PHP Fatal error:  Uncaught Aws\S3\Exception\MissingContentLengthException: AWS Error Code: MissingContentLength, Status Code: 411, AWS Request ID: ZESZXB7QM931H7F5, AWS Error Type: client, AWS Error Message: You must provide the Content-Length HTTP header., User-Agent: aws-sdk-php2/2.7.21 Guzzle/3.9.2 curl/7.36.0 PHP/5.5.21
  thrown in phar:///usr/syno/aws/aws.phar/Aws/Common/Exception/NamespaceExceptionFactory.php on line 91
```

**Comment by mtdowling** [comment]
Thanks for the additional information. I was able to reproduce this and found that the cause of the error was in Guzzle. I've pushed a fix for Guzzle (v3.9.3), and we will have this version bundled in the next release of the SDK's phar and zip files.
